agent/yubiagent/slot.go

```diff
@@ -0,0 +1,196 @@
+func (s *Slot) CSRs() []*proto.SSHCertificateSigningRequest {
+	return s.csrs
+}
+
+// AddCertsToAgent adds the certificates to the agent of the key slot.
+// The function implements the interface csr.AgentKey.
+// When the signer returns the signed certificates back to gensign, gensign could simply invoke the function
+// to add certificates to the corresponding agent.
+func (s *Slot) AddCertsToAgent(certs []ssh.PublicKey, comments []string) error {
+	for i, c := range certs {
+		cert, err := keyutil.CastSSHPublicKeyToCertificate(c)
+		if err != nil {
+			continue
+		}
+		err = s.yubiAgent.AddHardCert(cert, comments[i])
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// PublicKey returns the public key of the key slot.
+func (s *Slot) PublicKey() ssh.PublicKey {
+	return s.public
+}
+
+// TouchPolicy returns the touch policy of the key slot.
+func (s *Slot) TouchPolicy() keyid.TouchPolicy {
+	return s.policy
+}
+
+// Serial returns the serial number of the yubikey.
+func (s *Slot) Serial() (string, error) {
+	return yubiattest.ModHex(s.attest)
+}
+
+// SlotCode returns the code number of the key slot.
+func (s *Slot) SlotCode() string {
+	return s.code
+}
+
+// AttestCert returns the attest cert of the key slot.
+func (s *Slot) AttestCert() *x509.Certificate {
+	return s.attest
+}
+
+// Agent returns the yubiagent.
+func (s *Slot) Agent() YubiAgent {
+	return s.yubiAgent
+}
+
+// getTouchPolicy returns the touch policy coded in the given attestation certificate
+func getTouchPolicy(attestCert *x509.Certificate) keyid.TouchPolicy {
+	var touch = keyid.DefaultTouch
+	if attestCert == nil {
+		return touch
+	}
+	for _, ext := range attestCert.Extensions {
+		// NOTE: The following id is the touch policy stored in attestation certificate.
+		//       Refer: https://developers.yubico.com/PIV/Introduction/PIV_attestation.html
+		if ext.Id.String() == "1.3.6.1.4.1.41482.3.8" {
+			touch = keyid.TouchPolicy(ext.Value[1])
+		}
+	}
+	return touch
+}
+
+// YubiKeys with firmaware version 4.2.6 - 4.3.4 are only affected by
+// Infineon RSA key generation issue
+func isCVE201715361AffectedVersion(extensions []pkix.Extension) bool {
+	for _, ext := range extensions {
+		if ext.Id.String() == "1.3.6.1.4.1.41482.3.3" {
+			version := int(ext.Value[0])*100 + int(ext.Value[1])*10 + int(ext.Value[2])
+			if version >= 426 && version <= 434 {
+				log.Printf("affected firmware detected, %v", version)
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// validateYubikeyCrytpoAlgorithm validates whether the crypto algorithm is supported by YubiKey.
+func validateYubikeyCrytpoAlgorithm(pub crypto.PublicKey) bool {
+	switch pk := pub.(type) {
+	case *rsa.PublicKey:
+		switch pk.Size() * 8 {
+		case 2048:
+			return true
+		default:
+			return false
+		}
+	case *ecdsa.PublicKey:
+		switch pk.Curve {
+		case elliptic.P256(), elliptic.P384():
+			return true
+		default:
+			return false
+		}
+	// ed25519 is non-supported key algo for now.
```

Comment:
`ed25519` is supported by yubikey but ysshra doesn't yet support it, correct? If so, can we add a todo here?

Reply:
According to https://support.yubico.com/hc/en-us/articles/360013647820-YubiKey-4C, it is not supported in PIV mode.